### DISCUSSION TITLE
Brainstorming: Proxy Mode UX for Mitmweb

### DISCUSSION THREAD

**Discussion by mhils** [mitmproxy/mitmproxy#6868] [Ideas]
The first project @lups2000 will be tackling for GSoC is adding a new tab in mitmweb to enable users to configure the proxy mode from within mitmweb. Basically we want to transform the mitmweb top bar from this:

<img width="379" alt="image" src="https://github.com/mitmproxy/mitmproxy/assets/1019198/a9dec2c9-c409-4da4-9aea-2552ca14abab">

to something like this:

<img width="473" alt="image" src="https://github.com/mitmproxy/mitmproxy/assets/1019198/8136f9a1-dbe1-4517-9222-525c30028005">

The idea here is that when someone opens mitmweb for the first time, they are presented with a screen where they can dynamically turn on local redirect mode, WireGuard mode, spawn an additional reverse proxy, etc. A major driving factor for this is to improve the onboarding experience.

### User Stories

We plan to do some mockups for the proxy mode tab, but maybe it's useful to collect some requirements / user stories first. I'll start with some ideas, **please everyone chime in!**

1. As a new mitmproxy user, I want guidance on how to inspect the traffic I'm interested in.
  a) As a user who is interested in a local application, I want an easy path towards intercepting that traffic.
  b) As a user who is interested in a device's traffic, I want an easy path towards intercepting that traffic.
2. As a mitmproxy user, I want to capture only specific applications with local redirect mode.
4. As a mitmproxy user, I want to start a WireGuard server.
6. As a mitmproxy user, I want to add additional reverse proxies with a custom destination.
7. As a mitmproxy user, I want to add additional upstream proxies with a custom destination.
8. As a mitmproxy user, I want to remove running servers.
9. As a mitmproxy user, I want to see startup/configuration failures for proxy modes.
8. As a experienced mitmproxy user, I want to get straight to the flow list if there is traffic coming in.

Nice to have, but probably not required:

8. As a mitmproxy user, I want to add additional HTTP proxies.
9. As a mitmproxy user, I want to start a DNS server.
10. As a mitmproxy user, I want to specify a custom listen host and port per proxy mode.
11. As an advanced mitmproxy user, I want to start transparent mode servers.
12. As a mitmproxy user with a WireGuard setup, I want to specify a custom WireGuard configuration.

**Comment by rosydawn6** [comment]
For #3  Making the language more consistent - As a mitmproxy user, I want to add a WireGuard server. -  

I think #7 - As a mitmproxy user, I want to see startup/configuration failures for proxy modes. - (and maybe others)  would have applicability for mitmdump as well.

Some of these are more advanced so maybe could be put in a different location.

Also I see the tab list went _from_ File | Start | Options _to_ File | Capture | View | Options
That may be confusing to some

**Comment by lups2000** [comment]
First Mockup for the new tab **Capture**, let me know what you like and what you don't. Any thoughts? 

![image](https://github.com/mitmproxy/mitmproxy/assets/100372313/6adf5c3a-8486-4d7a-97ca-a3098c257b22)

**Reply by emanuele-em** [reply]
I like it! As the only thing I'd make the difference between _Current Setting_ and _Capture more traffic_, maybe with a color, otherwise they all seem to be "selectable"

**Comment by lups2000** [comment]
another option we can adopt: 
![image](https://github.com/mitmproxy/mitmproxy/assets/100372313/459d7456-6dc0-4bcf-8951-f7f72007bff1)

A problem with the previous one was that if we changed the mode on-the-fly, the order of the list changed. In this way, we can maintain the order and just apply different colors to indicate when a mode is **activated** and when it is **simply selected**. Any opinions?

**Reply by emanuele-em** [reply]
this version is my favorite one

**Reply by lups2000** [reply]
probably mine as well 😄

**Comment by mhils** [comment]
I'm not sure if I like the list of big buttons too much. How do you feel about something like this (also in Figma)?

<img width="543" alt="image" src="https://github.com/mitmproxy/mitmproxy/assets/1019198/663792c2-c9aa-4892-af0d-6cc5af32f935">

Maybe we can add illustrative icons or something like that, that'd make it really nice.

**Reply by emanuele-em** [reply]
I like it but, maybe, with this solutions is not clear that some options are mutually exclusive

**Reply by mhils** [reply]
Can you clarify what you mean by "some options are mutually exclusive"?

**Comment by roniemartinez** [comment]
Probably just some source of inspiration for mitmproxy. I've been using HTTP Toolkit for a while and I really like that you have more options.
<img width="1552" alt="Screenshot 2024-06-14 at 8 16 40 PM" src="https://github.com/mitmproxy/mitmproxy/assets/2573537/fe073930-13e6-47cc-9b81-66d70c9653d1">
<img width="1552" alt="Screenshot 2024-06-14 at 8 19 51 PM" src="https://github.com/mitmproxy/mitmproxy/assets/2573537/e1f534b9-3028-4570-ad06-58d40863d838">

**Reply by lups2000** [reply]
Hi,
Thank you for your feedback, it's really appreciated! 👍 
We decided to stick with the latest mockup proposed by @mhils, at least for now. Our primary goal is to keep the UI simple and intuitive to enhance the average user experience. Additionally, some of our modes require extra settings, and using big buttons would necessitate a sidebar, making the UI more complicated 😄

**Reply by mhils** [reply]
@roniemartinez: Thanks! This actually brings up some interesting ideas. I think we totally could add a "Start a fresh preconfigured Chrome instance" button right below the "Explicit HTTP(S) Proxy" configuration in a future step. :) We do have the backend functionality for that already, so it would be easy to implement.
